File: General-DSA-Questions/Heap/11_mergeKSortedLinkedList.cpp
// Write a function to merge k sorted linked lists into a single sorted linked list.

// This is similar to merging k sorted arrays.

// Approach:
// 1. Use a min-heap that compares nodes based on their value.
// 2. Push the head of each linked list into the heap.
// 3. While the heap is not empty:
//    a.  Pop the smallest node.
//    b. Add it to the merged list.
//    c.  If the popped node has a next, push it into the heap.
// 4. Return the merged linked list.
// 5. Time Complexity: O(N log K)
//    a.  Where N = total number of nodes, K = number of lists
// 6. Space Complexity: O(K) for the heap

#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Definition of a linked list node
class Node {
public:
    int data;
    Node* next;

    Node(int val) {
        data = val;
        next = nullptr;
    }
};

// Custom comparator for the min-heap
struct Compare {
    bool operator()(Node* a, Node* b) {
      // This says:
      //   Return true if a is greater than b.
      //   That tells the priority_queue that a should go after b.
      //   So, smaller values will come first, forming a min-heap.
        return a->data > b->data; // Compare based on node values
    }
};

// Function to merge k sorted linked lists into a single sorted linked list
Node* mergeKSortedLists(vector<Node*>& lists) {
    // priority_queue<Type, Container, Comparator> name;
    priority_queue<Node*, vector<Node*>, Compare> minHeap; // Min-heap to store the nodes

    // Push the head of each linked list into the min-heap
    for (Node* list : lists) {
        if (list != nullptr) {
            minHeap.push(list);
        }
    }

    Node* dummy = new Node(0); // Dummy node to simplify the merged list
    Node* tail = dummy; // Pointer to the last node in the merged list

    // While the min-heap is not empty
    while (!minHeap.empty()) {
        Node* current = minHeap.top(); // Get the smallest node
        minHeap.pop(); // Remove it from the heap

        tail->next = current; // Add it to the merged list
        tail = tail->next; // Move to the last node in the merged list

        // If there is a next node in the same linked list, push it into the min-heap
        if (current->next != nullptr) {
            minHeap.push(current->next);
        }
    }

    tail->next = nullptr; // Set the next of the last node to null

    return dummy->next; // Return the merged linked list (skipping the dummy node)
}

// Function to print the linked list
void printList(Node* head) {
    while (head != nullptr) {
        cout << head->data << " -> ";
        head = head->next;
    }
    cout << "nullptr" << endl;
}

// Function to create list of k sorted linked lists
vector<Node*> createSortedLists() {
    vector<Node*> lists;

    // Create first sorted linked list: 1 -> 4 -> 7
    Node* list1 = new Node(1);
    list1->next = new Node(4);
    list1->next->next = new Node(7);
    lists.push_back(list1);

    // Create second sorted linked list: 2 -> 5 -> 8
    Node* list2 = new Node(2);
    list2->next = new Node(5);
    list2->next->next = new Node(8);
    lists.push_back(list2);

    // Create third sorted linked list: 3 -> 6 -> 9
    Node* list3 = new Node(3);
    list3->next = new Node(6);
    list3->next->next = new Node(9);
    lists.push_back(list3);

    return lists;
}

int main() {
    vector<Node*> lists = createSortedLists(); // Create k sorted linked lists

    Node* mergedList = mergeKSortedLists(lists); // Merge the k sorted linked lists

    cout << "Merged sorted linked list: ";
    printList(mergedList); // Print the merged linked list

    return 0;
}

// Output:
// Merged sorted linked list: 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> 8 -> 9 -> nullptr

File: JS-DSA-Questions/Heap/11_mergeKSortedLinkedList.js
// Merge k sorted linked lists into a single sorted linked list.

// Approach (Min-Heap):
// 1. Push the head node of each list into a min-heap ordered by node value.
// 2. Pop the smallest node, append it to the merged list, and push its next node (if any).
// 3. Repeat until the heap is empty.
// Time: O(N log K) (N total nodes, K lists). Space: O(K).

class Node {
  constructor(val) {
    this.data = val;
    this.next = null;
  }
}

// JavaScript note: C++ used a priority_queue with a custom Compare on Node*. Here we include a
// comparator-based MinHeap and order by node.data.
class MinHeap {
  constructor(compare) {
    this.a = [];
    this.compare = compare;
  }
  size() {
    return this.a.length;
  }
  push(v) {
    this.a.push(v);
    let i = this.a.length - 1;
    while (i > 0) {
      const p = Math.floor((i - 1) / 2);
      if (this.compare(this.a[p], this.a[i]) > 0) {
        [this.a[p], this.a[i]] = [this.a[i], this.a[p]];
        i = p;
      } else break;
    }
  }
  pop() {
    const top = this.a[0];
    const last = this.a.pop();
    if (this.a.length > 0) {
      this.a[0] = last;
      let i = 0;
      const n = this.a.length;
      while (true) {
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        let s = i;
        if (l < n && this.compare(this.a[l], this.a[s]) < 0) s = l;
        if (r < n && this.compare(this.a[r], this.a[s]) < 0) s = r;
        if (s !== i) {
          [this.a[i], this.a[s]] = [this.a[s], this.a[i]];
          i = s;
        } else break;
      }
    }
    return top;
  }
}

function mergeKSortedLists(lists) {
  const minHeap = new MinHeap((x, y) => x.data - y.data);

  for (const list of lists) {
    if (list !== null) {
      minHeap.push(list);
    }
  }

  const dummy = new Node(0);
  let tail = dummy;

  while (minHeap.size() > 0) {
    const current = minHeap.pop();
    tail.next = current;
    tail = tail.next;

    if (current.next !== null) {
      minHeap.push(current.next);
    }
  }

  tail.next = null;
  return dummy.next;
}

function printList(head) {
  const out = [];
  while (head !== null) {
    out.push(head.data);
    head = head.next;
  }
  console.log(out.join(" -> ") + " -> null");
}

function createSortedLists() {
  const lists = [];
  const list1 = new Node(1);
  list1.next = new Node(4);
  list1.next.next = new Node(7);
  lists.push(list1);

  const list2 = new Node(2);
  list2.next = new Node(5);
  list2.next.next = new Node(8);
  lists.push(list2);

  const list3 = new Node(3);
  list3.next = new Node(6);
  list3.next.next = new Node(9);
  lists.push(list3);

  return lists;
}

const merged = mergeKSortedLists(createSortedLists());
process.stdout.write("Merged sorted linked list: ");
printList(merged); // 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> 8 -> 9 -> null
